new features for interactions and block structures

the-hampel opened the pull request:
* enforce PEP standards in interaction_hamiltonian.py
* allow different interaction Hamiltonians per impurity
* add new interaction of format U/2 * (Ntot^2 - Ntot)
* add pick solver struct option to pick specific blocks to be treated in the impurity solver

the-hampel committed:
enforce PEP standard in interaction Hamiltonian

the-hampel committed:
allow different interaction Hamiltonians per impurity

the-hampel committed:
add Ntot interaction

the-hampel committed:
fix print for failing comparison test

the-hampel committed:
add pick solver struct option to select specific blocks for the impurity problem

the-hampel committed:
fix header writing for obs file

the-hampel committed:
aplly block picker before block mapping

the-hampel committed:
function to det n_orb_solver, fix test

merkelm committed:
Few small bug fixes

merkelm reviewed (commented):
Thanks Alex for the pull request, it will be very helpful. I added a few changes and small bug fixes in my commit here, feel free to ask me about/undo anything you disagree with.

A few points I'd like to mention separately:
- Is solver.py the best location for n_orb_solver (or solver.get_n_orbitals in my suggestion)? It's a stateless function acting on the sumk object
- Did you convince yourself that pick_solver_struct implements something that map_solver_struct cannot do?
- The `ratio_F2_F4` is now always used which is needed because of the list of interaction Hamiltonians, right? It would be nice to check with `valid for` option that it is not set for Hamiltonians that don't support it

the-hampel committed:
include suggestions of @merkelm

* improve doc string of pick and map gf_struct_solver
* add valid_for condition for ratio_F4_F2
* map_gf_struct_solver must now be a list for each imp
* check that pick and map_gf_struct_solver are a list

the-hampel committed:
fix parameter n_orb in hint construction

when pick_gf_struct_solver or map_gf_struct solver is used, the number
of orbital n_orb of sum_k.corr_shell[ish].dim is different from the
actual solver space dimension. Use function get_n_orbitals to determine
correct number of orbitals.

the-hampel commented:
Hi Max, 
I did the requested changes and tested them. I had to fix one problem in [#babd79d ](https://github.com/TRIQS/solid_dmft/pull/27/commits/babd79db0ceb41c2e5eaa6e86f939d82d6cef069) to fix how the number of orbitals is determined. This was failing when pick_gf_struct_solver is used to reduce the size of the impurity space. Please check if that makes sense to you. 

Also I am happy to discuss the location of the function `get_n_orbitals`. Just felt that calling something like solver.get_n_orbitals sound good. But we can rename and put it somewhere else. Any suggestions?
